File: src/update_heading.py
from std_msgs.msg import Float64

def update_heading(depth_control_signal, heading_control_signal, stability_thrust, rpm_adjustment):

    # Define maximum RPM limits
    max_rpm_depth = 400
    heading_control_signal = max(min(heading_control_signal, 25), -25)

    # Apply constraints to thruster commands to be within +/- 1200 RPM
    vert_port_command = max(min(-depth_control_signal, max_rpm_depth), -max_rpm_depth)  # Positive RPMs send down
    vert_stbd_command = max(min(depth_control_signal, max_rpm_depth), -max_rpm_depth)  # Positive RPMs send up

    bow_port_command = -stability_thrust + heading_control_signal  # Negative for clockwise due to positive RPMs send CCW
    bow_stbd_command = stability_thrust + heading_control_signal   # Positive sends CCW, need negative for CW
    aft_port_command = -stability_thrust + rpm_adjustment
    aft_stbd_command = stability_thrust - rpm_adjustment

    return {
        'vert_port': Float64(data=vert_port_command),
        'vert_stbd': Float64(data=vert_stbd_command),
        'bow_port': Float64(data=bow_port_command),
        'bow_stbd': Float64(data=bow_stbd_command),
        'aft_port': Float64(data=aft_port_command),
        'aft_stbd': Float64(data=aft_stbd_command)
    }

File: src/test_scripts/Young_Project2/updateHeading.m
function [vertPortThrusterMsg, vertStbdThrusterMsg, ...
    bowPortThrusterMsg, bowStbdThrusterMsg, ...
    aftPortThrusterMsg, aftStbdThrusterMsg, aftVertThrusterMsg] = ...
    updateHeading(depthControlSignal, stabilityThrust, headingControlSignal, rpmAdjustment,... 
    vertPortThrusterMsg, vertStbdThrusterMsg, ...
    bowPortThrusterMsg, bowStbdThrusterMsg,...
    aftPortThrusterMsg, aftStbdThrusterMsg, aftVertThrusterMsg)

    % assign maximum RPM limits
    MaxRPM_Depth = 400;
    headingControlSignal = max(min(headingControlSignal, 25), -25);

    % Apply constraints to thruster commands to be within +/- 1200 RPM
    vertPortCommand = max(min(-depthControlSignal , MaxRPM_Depth), -MaxRPM_Depth);  % Positive RPMs send down
    vertStbdCommand = max(min(depthControlSignal, MaxRPM_Depth), -MaxRPM_Depth);  % Positive RPMs send up

    bowPortCommand = -stabilityThrust + headingControlSignal;  % Negative for clockwise due to positive RPMs send CCW
    bowStbdCommand = stabilityThrust + headingControlSignal ;  % Positive sends CCW, need negative for CW
    aftPortCommand = -stabilityThrust + rpmAdjustment;
    aftStbdCommand = stabilityThrust - rpmAdjustment;

    %aftVertCommand = max(min(pitchControlSignal, MaxRPM_rotational), -MaxRPM_rotational);

    % Set the command in the message data
    vertPortThrusterMsg.Data = vertPortCommand;
    vertStbdThrusterMsg.Data = vertStbdCommand;
    bowPortThrusterMsg.Data = bowPortCommand; 
    bowStbdThrusterMsg.Data = bowStbdCommand;
    aftPortThrusterMsg.Data = aftPortCommand;
    aftStbdThrusterMsg.Data = aftStbdCommand;
   % aftVertThrusterMsg.Data = aftVertCommand;
end

